### PR TITLE
Location puck velocity fix.

### PR DESCRIPTION
**Pull request by tomaszrybakiewicz**
Closes [#4140](https://github.com/mapbox/mapbox-navigation-android/issues/4140)

### Description

**tl;dr;** Using PathInterpolator to animate location puck at a constant speed over key points.

- Introduced `ConstantVelocityInterpolator` that uses Android PathInterpolator to calculate timings for ValueAnimator with key points.
- Added `PuckAnimationEvaluator` encapsulates wiring of `ConstantVelocityInterpolator` with location puck ValueAnimators.
- Updated `NavigationLocationProvider` to use `PuckAnimationEvaluator` to configure puck position ValueAnimator.

### Screenshots or Gifs

**Before**

https://user-images.githubusercontent.com/2678039/173482267-01fef844-cfc6-4571-a0bd-9c6ca2759eda.mp4

**After**

https://user-images.githubusercontent.com/2678039/173482098-26aa5099-4944-41c2-a15f-8255e3c4361a.mp4

### CODE REVIEW

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        if (0 < total) {
+            val path = Path()
+            val step = 1.0f / keyPoints.size
+            var pathTime = 0.0
+            keyPoints.forEachIndexed { index, point ->
+                val deltaTime = distances[index] / total
+                pathTime += deltaTime
+                path.lineTo(pathTime.toFloat(), step * (index + 1))
+            }
+            innerInterpolator = PathInterpolator(path)
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt:38]: It can be val if you add an else branch here.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+
+    override fun getInterpolation(input: Float): Float =
+        innerInterpolator.getInterpolation(input)
+
+    private fun distanceSqrTo(p1: Point, p2: Point): Double {
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt:45]: Maybe move this to some Utils?

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        val sut = ConstantVelocityInterpolator(p0, keyPoints)
+
+        val path = calculatePathValues(p0, *keyPoints)
+        assertEquals(keyPoints.size, path.size)
+        path.forEach { (outTime, inTime) ->
+            assertEquals(outTime, sut.getInterpolation(inTime))
+        }
+    }
+
+    private fun calculatePathValues(p0: Point, vararg points: Point): List<Pair<Float, Float>> {
```
Comment (dzinad) [inline libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:31]: Is it just the copypaste of ConstantVelocityInterpolator init? Is there really much sense in such a test? I'd say we should take a simple example, calculate the expected value manually and check it in the test.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.Assert.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import kotlin.math.pow
+
+@RunWith(RobolectricTestRunner::class)
+class ConstantVelocityInterpolatorTest {
+
+    @Test
```
Comment (dzinad) [inline libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:13]: There are no tests for:
1) An empty array;
2) Array with single point;
3) Array with 2 identical points (so that the distance is zero).

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomaszrybakiewicz) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt:45]: There is no need to move this to utils. This is only needed for this class.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt (hunk 3 above)
Comment (tomaszrybakiewicz) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:31]: This test not only validates PathInterpolator values but also verifies velocity is constant. Calculating this by hand and hardcoding the result here brings no value.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt (hunk 4 above)
Comment (korshaknn) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:13]: > There is no need for such tests.

why? shouldn't we handle such edge cases and have tests for them?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt (hunk 1 above)
Comment (tomaszrybakiewicz) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt:38]: Great suggestion. Thx.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/PuckAnimationEvaluator.kt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    private var interpolator: TimeInterpolator? = null
+
+    override fun getInterpolation(input: Float): Float =
+        interpolator?.getInterpolation(input) ?: input
+
+    override fun evaluate(fraction: Float, startValue: Point, endValue: Point): Point {
+        if (interpolator == null) {
+            // we defer creation of TimeInterpolator until we know startValue
+            interpolator = ConstantVelocityInterpolator(startValue, keyPoints)
```
[the remark this answers is not in the document]
Comment (tomaszrybakiewicz) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/PuckAnimationEvaluator.kt:26]: This class is meant to be instantiated once for each ValueAnimator. But I see what you mean. Someone might try to re-use it multiple times.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt (hunk 3 above)
Comment (dzinad) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:31]: Why no value? To be more precise, why does it bring less value than copypasting the code what you test? If you've made an error in the implementation it will also be present here. What's the point of the test then?

> This test not only validates PathInterpolator values but also verifies velocity is constant.

If this causes some issues it can be divided into different tests.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt (hunk 4 above)
Comment (dzinad) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:13]: I also don't see why we shouldn't test it.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt (hunk 3 above)
Comment (tomaszrybakiewicz) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:31]: I know it can, but there is no need for it. `calculatePathValues` method already does that for us. It performs a self-check to guarantee that returned "verfication path" does produce timing for constant velocity.
With that being guaranteed, we can use that "verification path" to validate  ConstantVelocityInterpolator values.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt (hunk 4 above)
Comment (tomaszrybakiewicz) [reply libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolatorTest.kt:13]: Done. Added more tests that verify above cases.

**Comment by tomaszrybakiewicz** [comment]
@LukasPaczos as you can see in the second video from the description, there is still some jitter and small lag when animating puck position using key points. I ran out of ideas about what could be causing it. WDYT?

**Review by dzinad — APPROVED** [review]

**Comment by Guardiola31337** [comment]
Noting that `routeRefreshesWorksAfterSettingsNewRoutes` failed 👀 

```
06-14 10:28:19.685: E/TestRunner(2501): failed: routeRefreshesWorksAfterSettingsNewRoutes(com.mapbox.navigation.instrumentation_tests.core.RouteRefreshTest)
06-14 10:28:19.685: E/TestRunner(2501): ----- begin exception -----
06-14 10:28:19.690: E/TestRunner(2501): kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 5000 ms
06-14 10:28:19.690: E/TestRunner(2501): 	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:186)
06-14 10:28:19.690: E/TestRunner(2501): 	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:156)
06-14 10:28:19.690: E/TestRunner(2501): 	at android.os.Handler.handleCallback(Handler.java:739)
06-14 10:28:19.690: E/TestRunner(2501): 	at android.os.Handler.dispatchMessage(Handler.java:95)
06-14 10:28:19.690: E/TestRunner(2501): 	at android.os.Looper.loop(Looper.java:148)
06-14 10:28:19.690: E/TestRunner(2501): 	at android.app.ActivityThread.main(ActivityThread.java:5417)
06-14 10:28:19.690: E/TestRunner(2501): 	at java.lang.reflect.Method.invoke(Native Method)
06-14 10:28:19.690: E/TestRunner(2501): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
06-14 10:28:19.690: E/TestRunner(2501): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
06-14 10:28:19.690: E/TestRunner(2501): ----- end exception -----
```

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            TimeInterpolator { it }
+        }
+    }
+
+    override fun getInterpolation(input: Float): Float =
+        innerInterpolator.getInterpolation(input)
+
+    private fun distanceSqrTo(p1: Point, p2: Point): Double {
+        return (p2.latitude() - p1.latitude()).pow(2) +
+            (p2.longitude() - p1.longitude()).pow(2)
```
Comment (LukasPaczos) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt:49]: This doesn't look correct, shouldn't it be
```suggestion
        return sqrt((p2.latitude() - p1.latitude()).pow(2) +
            (p2.longitude() - p1.longitude()).pow(2))
```
or basically
```suggestion
        return hypot(p2.latitude() - p1.latitude(), p2.longitude() - p1.longitude())
```

Comment (tomaszrybakiewicz) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt]: Oh @#$%! You are right!!!! My math was wrong. This fixed it. Now works as expected.

https://user-images.githubusercontent.com/2678039/173883490-b525e528-4bbf-44c2-ab16-e302ded44b67.mp4

.

Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt]: Nice, this looks so good right now!

Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/location/ConstantVelocityInterpolator.kt]: Great work putting this together.

**Comment by LukasPaczos** [comment]
Could you add a changelog entry as well? It'd be great to highlight this improvement.

**Comment by abhishek1508** [comment]
Testing the branch locally and setting replay speed to `5.0` in `MapboxCameraAnimationsActivity`, I observe the following:

#### On main branch
https://user-images.githubusercontent.com/9770186/173894874-5e8f1ec2-fb56-4a1d-a4b4-9531ea055817.mp4

#### On this branch
https://user-images.githubusercontent.com/9770186/173894946-bec542cb-ca8c-4652-86db-2566dbde8382.mp4

On both the branches I see the puck cutting corners, but it's a lot on this branch. Given that this is a simulator based testing, I am not sure if we should be concerned here. 

/cc @Guardiola31337 @LukasPaczos @tomaszrybakiewicz

**Comment by tomaszrybakiewicz** [comment]
> On both the branches I see the puck cutting corners, but it's a lot on this branch. Given that this is a simulator based testing, I am not sure if we should be concerned here.

@abhishek1508 I don't think the time interpolation method used here causes this. 

I think it's the mismatch between animation duration and location updates frequency from the replayer. Notice, in `MapboxCameraAnimationsActivity.locationObserver` we we are calling `navigationLocationProvider.changePosition()` with animation `duration` set to 1000ms in order to match NavigationCamera animation speed. 

If set replayer speed to 5x, we willl receive each location update at 5x speed as well. If we call `navigationLocationProvider.changePosition()`, without changing anim duration, the Puck's ValueAnimator will only animate for 200ms before being canceled and replaced by the next animator.

Here's a recording of a replayer with 5x speed and animation duration set to 1000/5 = 200ms.

https://user-images.githubusercontent.com/2678039/173899787-f1e4f694-dceb-4f78-a06f-6b1e1579354f.mp4

**Review by abhishek1508 — APPROVED** [review]
Tested it thoroughly and this is looking pretty good. Great work 🎉

**Comment by tomaszrybakiewicz** [comment]
_static-analysis_ check failed. Re-running.
